src/plugins_types_bits.c

```diff
@@ -62,113 +62,120 @@ ly_type_store_bits(const struct ly_ctx *ctx, const struct lysc_type *type, const
     ret = ly_set_new(&items);
     LY_CHECK_GOTO(ret != LY_SUCCESS, cleanup_value);
 
-    ws_count = lws_count = 0;
-    for (index = 0; index < value_len; ++index) {
-        if (isspace(value[index])) {
-            ++ws_count;
-            continue;
+    /* get all values */
+    while (index_end < value_len) {
+        /* skip leading spaces */
+        index = index_end;
+        while ((index < value_len) && isspace(value[index])) {
+            index++;
         }
-        if (index == ws_count) {
-            lws_count = ws_count;
-        } else if (ws_count) {
-            iscanonical = 0;
+
+        index_end = index;
+        /* find end of word */
+        while ((index_end < value_len) && !isspace(value[index_end])) {
+            index_end++;
         }
-        ws_count = 0;
 
-        /* start of the item */
+        /* check if name of bit is valid */
         item = &value[index];
-        for (item_len = 0; index + item_len < value_len && !isspace(item[item_len]); item_len++) {}
-        LY_ARRAY_FOR(type_bits->bits, u) {
-            if (!ly_strncmp(type_bits->bits[u].name, item, item_len)) {
-                /* we have the match */
-                uint32_t inserted;
-
-                if (iscanonical && items->count && (type_bits->bits[u].position < ((struct lysc_type_bitenum_item *)items->objs[items->count - 1])->position)) {
-                    iscanonical = 0;
-                }
-                ret = ly_set_add(items, &type_bits->bits[u], 0, &inserted);
-                LY_CHECK_GOTO(ret, cleanup);
-                if (inserted != items->count - 1) {
-                    ret = ly_err_new(err, LY_EVALID, LYVE_DATA, NULL, NULL,
-                            "Bit \"%s\" used multiple times.", type_bits->bits[u].name);
-                    goto cleanup;
-                }
-                goto next;
+        item_len = index_end - index;
+        if (item_len != 0) {
+            uint32_t item_pos;
+            uint32_t item_pos_expected;
+            LY_ARRAY_COUNT_TYPE array_it = 0;
+            LY_ARRAY_COUNT_TYPE array_size = LY_ARRAY_COUNT(type_bits->bits);
+
+            /* looking for correct name */
+            while ((array_it < array_size) &&
```

Comment:
I believe that using `LY_ARRAY_FOR` is a more common and therefore also a more clear way to go through the sized array. The good point is to avoid the `next` label, but I would stay more with `LY_ARRAY_FOR` instead of rewriting this part to use `while` loops.

Reply:
For me is more readable when code contains gotos as less as possible
 I'm convinced that command goto should be used only for error handling.

Reply:
I would not be so strict here and let everyone use whatever they prefer. As for `goto`, I agree it should preferably be avoided but what should strictly be followed is `goto` referencing only a label **below** itself. This is even what some tools checks and consider an error.

Reply:
As I wrote, I'm appreciating removing the goto here, and I don't want to be strict regarding the while/for, but since most of the libyang code use `LY_ARRAY_FOR` for this purpose, it seems more natural to stick with it also here instead of using while in this part. Or am I'm missing some important point?

Reply:
OK. Probably I can write it with FOR without goto.

Reply:
Thanks